tests/integration/integration_nodejs_test.go

```diff
@@ -1287,6 +1287,28 @@ func TestESMTS(t *testing.T) {
 	})
 }
 
+func TestESMTSAuto(t *testing.T) {
+	t.Parallel()
+	dir := filepath.Join("nodejs", "esm-ts-auto")
+	e := ptesting.NewEnvironment(t)
+	defer e.DeleteIfNotFailed()
+	e.ImportDirectory(dir)
+	stackName := ptesting.RandomStackName()
+	// For this test we need to properly install the core SDK instead of yarn
+	// linkining, because yarn link breaks finding an alternative ts-node
+	// installation. This would cause th test to fallback to the vendored
+	// ts-node, which does not provide ts-node/esm.
+	coreSDK, err := filepath.Abs(filepath.Join("..", "..", "sdk", "nodejs", "bin"))
+	require.NoError(t, err)
+	e.RunCommand("yarn", "install")
+	e.RunCommand("yarn", "add", coreSDK)
+	e.RunCommand("pulumi", "login", "--cloud-url", e.LocalURL())
+	e.RunCommand("pulumi", "stack", "init", stackName)
+	e.RunCommand("pulumi", "stack", "select", stackName)
+	e.RunCommand("pulumi", "up", "--yes")
+	e.RunCommand("pulumi", "destroy", "--skip-preview", "--refresh=true")
```

Comment:
Should we not be checking the stack exports are as expected?

Reply:
Good call, I added that in https://github.com/pulumi/pulumi/pull/19980/commits/97c0f02932c3b957896770d0684b1898c259ad0b

I also added a test that validates we get type checking errors. This is important because recent Node.js versions can directly load `.ts` files and just strip the types, not doing any type checking. We want to validate we still go through `ts-node` and get errors.